tls/s2n_handshake.c

```diff
@@ -100,3 +101,60 @@ int s2n_handshake_get_hash_state(struct s2n_connection *conn, s2n_hash_algorithm
     }
     return 0;
 }
+
+int s2n_handshake_require_all_hashes(struct s2n_connection *conn)
+{
+    memset(conn->handshake.required_hash_algs, 1, sizeof(conn->handshake.required_hash_algs));
+    return 0;
+}
+
+int s2n_handshake_require_hash(struct s2n_handshake *handshake, s2n_hash_algorithm hash_alg)
+{
+    handshake->required_hash_algs[hash_alg] = 1;
+    return 0;
+}
+
+uint8_t s2n_handshake_is_hash_required(struct s2n_handshake *handshake, s2n_hash_algorithm hash_alg)
+{
+    return handshake->required_hash_algs[hash_alg];
+}
+
+/* Update the required handshake hash algs depending on current handshake session state.
+ * This function must called at the end of a handshake message handler.
+ */
+int s2n_conn_update_required_handshake_hashes(struct s2n_connection *conn)
+{
+    /* Clear all of the required hashes */
+    memset(conn->handshake.required_hash_algs, 0, sizeof(conn->handshake.required_hash_algs));
+
+    /* Set the hash alg(s) required for the PRF */
+    switch (conn->actual_protocol_version) {
+    case S2N_SSLv3:
+    case S2N_TLS10:
+    case S2N_TLS11:
+        GUARD(s2n_handshake_require_hash(&conn->handshake, S2N_HASH_MD5));
+        GUARD(s2n_handshake_require_hash(&conn->handshake, S2N_HASH_SHA1));
+        break;
+    case S2N_TLS12:
+    {
+        /* For TLS 1.2 the cipher suite defines the PRF hash alg */
+        s2n_hmac_algorithm tls12_prf_alg = conn->secure.cipher_suite->tls12_prf_alg;
+        s2n_hash_algorithm hash_alg;
+        GUARD(s2n_hmac_hash_alg(tls12_prf_alg, &hash_alg));
+        GUARD(s2n_handshake_require_hash(&conn->handshake, hash_alg));
+        break;
+    }
+    }
+
+    message_type_t handshake_message = s2n_conn_get_current_message_type(conn);
+    const uint8_t client_cert_verify_done = (handshake_message >= CLIENT_CERT_VERIFY) ? 1 : 0;
+    s2n_cert_auth_type client_cert_auth_type;
+    GUARD(s2n_connection_get_client_auth_type(conn, &client_cert_auth_type));
+
+    /* If client authentication is possible, all hashes are needed until we're past CLIENT_CERT_VERIFY. */
+    if ((client_cert_auth_type != S2N_CERT_AUTH_NONE) && !client_cert_verify_done) {
```

Comment:
Should this Client Auth check be moved to the top of the function so that it can `return 0;` early?

Reply:
Good idea, that should reduce the # of non terminal branches.